packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -38,19 +38,31 @@ export default createRule({
       const checker = parserServices.program.getTypeChecker();
       let counts: Map<ts.Identifier, number> | undefined;
 
+      // Get the scope in which the type parameters are declared.
+      const scope = context.sourceCode.getScope(node);
+
       for (const typeParameter of tsNode.typeParameters) {
         const esTypeParameter =
           parserServices.tsNodeToESTreeNodeMap.get<TSESTree.TSTypeParameter>(
             typeParameter,
           );
-        const scope = context.sourceCode.getScope(esTypeParameter);
+
+        const smTypeParameterVariable = nullThrows(
+          scope.variables.find(
+            variable =>
+              // type identifiers can only be declared once, unlike var.
+              variable.identifiers.length === 1 &&
+              variable.identifiers[0] === esTypeParameter.name,
+          ),
```

Comment:
You can use `scope.set.get(esTypeParameter.name)` - there's a map storing the vars by name.

Reply:
Also you can use `variable.isTypeVariable` to see if it's a type definition. If you want to make doubly sure also check `!variable.isValueVariable`

Reply:
made changes 👍